### PR TITLE
Kinase-informed models

### PR DESCRIPTION
**Pull request by t-kimber**
## Description
Fix the kinase-informed models.

## Todos
  - [x] Fix input shape

## Status
- [x] Ready to go

### CODE REVIEW

File: kinoml/ml/torch_models.py [hunk 1]
```diff
@@ -179,28 +183,182 @@ def __init__(
...
-        self.temp = (self.max_length - self.kernel_shape + 1) * self.embedding_shape
+        self.temp = (self.input_shape[1] - self.kernel_shape + 1) * self.embedding_shape
         self.fully_connected_1 = nn.Linear(self.temp, self.hidden_shape)
         self.fully_connected_out = nn.Linear(self.hidden_shape, self.output_shape)
 
     def forward(self, x):
         """
         Defines the foward pass for a given input 'x'
         """
+        x = x[0]
```
Comment (jchodera) [inline kinoml/ml/torch_models.py:206]: Can we document what shape is expected for `x` as input, and why this line (`x = x[0]`) is appropriate here?

File: kinoml/ml/torch_models.py [hunk 2]
```diff
@@ -179,28 +183,182 @@ def __init__(
...
+        )
+        self.temp_protein = (self.input_shape[1][1] - self.kernel_shape + 1) * self.embedding_shape
+
+        self.fully_connected_1 = nn.Linear(self.temp_ligand + self.temp_protein, self.hidden_shape)
+        self.fully_connected_out = nn.Linear(self.hidden_shape, self.output_shape)
+
+    def forward(self, x):
+        """
+        Defines the foward pass for given an input composed of two entities: ligand and protein.
+        """
```
Comment (jchodera) [inline kinoml/ml/torch_models.py:282]: Can we document the expected shape and content of `x` here?

File: kinoml/ml/torch_models.py [hunk 3]
```diff
@@ -179,28 +183,182 @@ def __init__(
...
+       self.ligand_protein = self.embedding_shape_ligand + self.embedding_shape_protein
+
+       # Output
+       self.fully_connected_1 = nn.Linear(self.ligand_protein, self.hidden_shape)
+       self.fully_connected_out = nn.Linear(self.hidden_shape, self.output_shape)
+
+   def forward(self, x):
+       """
+       Defines the foward pass for given an input composed of two entities: ligand and protein.
+       """
```
Comment (jchodera) [inline kinoml/ml/torch_models.py:355]: Can we document the expected shape and content of `x` in the docstring?

File: kinoml/ml/torch_models.py [hunk 4]
```diff
@@ -22,11 +22,16 @@ def estimate_input_shape(input_sample):
         If your ``.forward()`` method takes something else
         than a Tensor, please adapt this method accordingly.
         """
-        if type(input_sample) == list:
-            return input_sample[0].shape[1]
-        else:
-            return input_sample.shape[1]
-
+        if len(input_sample)==1: # ligand-only
```
Comment (jchodera) [inline kinoml/ml/torch_models.py:25]: I'm a bit confused why `input_sample` has such complex structure that requires all the special cases here.
Can we document the convention we're using for storing data in `input_sample` a bit more thoroughly, or point to where that is documented if it's documented already?

**Review by jchodera — APPROVED** [review]
Looks like this works, but I've pointed out a few places where a bit of extra documentation can improve readability in terms of understanding how data is structured and unpacked.

File: kinoml/ml/torch_models.py (hunk 4 above)
Comment (schallerdavid) [reply kinoml/ml/torch_models.py:25]: Will those if statements catch all input_sample formats that may show up in future?

File: kinoml/ml/torch_models.py [hunk 5]
```diff
@@ -165,12 +170,11 @@ class ConvolutionNeuralNetworkRegression(_BaseModule):
         The activation function used in the hidden (only!) layer of the network.
     """
 
-    needs_input_shape = False
+    needs_input_shape = True
```
Comment (schallerdavid) [inline kinoml/ml/torch_models.py:173]: Is this variable actually being used?

File: kinoml/ml/torch_models.py [hunk 6]
```diff
@@ -179,28 +183,182 @@ def __init__(
...
         Defines the foward pass for a given input 'x'
         """
+        x = x[0]
+        x = x.float()
         x = self._activation(self.convolution(x))
         x = torch.flatten(x, 1)
         x = self._activation(self.fully_connected_1(x))
         return self.fully_connected_out(x)
+
+class ConvolutionNeuralNetworkRegressionKinaseInformed(_BaseModule):
```
Comment (schallerdavid) [inline kinoml/ml/torch_models.py:213]: Just wondering about the name: Of course KinoML is kinase-centric right now. But it was always designed to be generalizable to other protein families. Wouldn't this class also work on other protein families?

File: kinoml/ml/torch_models.py [hunk 7]
```diff
@@ -179,28 +183,182 @@ def __init__(
...
+        x_lig = torch.flatten(x_lig, 1)
+        x_prot = torch.flatten(x_prot, 1)
+
+        x = torch.cat((x_lig, x_prot), dim=1)
+
+        x = self._activation(self.fully_connected_1(x))
+
+        return self.fully_connected_out(x)
+
+class NeuralNetworkRegressionKinaseInformed(_BaseModule):
```
Comment (schallerdavid) [inline kinoml/ml/torch_models.py:297]: Same as above, would ProteinInformed also work here as a name?

File: kinoml/ml/torch_models.py (hunk 1 above)
Comment (t-kimber) [reply kinoml/ml/torch_models.py]: The reason why we access the first (or 2 first, or more) elements of `x` is due to the way the parquet files are constructed.
I added details and hope it makes it clearer.

Here's the documentation link: 
https://github.com/openkinome/experiments-binding-affinity/blob/master/features/featurize-template.ipynb , 
and paragraph:



Write tensors to disk
Output files are written to _output/<PIPELINE>/<DATASET>/<GROUP>.parquet files.
Each parquet will contain at least two array-like objects. The dimensionality of the parquet files is built as (systems, X_or_y, ...). For example, the first X vector for the first system is accessed like parquet[0, "0"]. Notice how the 2nd index is a string! (awkward design).

"0" (X, featurized systems). See DatasetProvider.to_awkward for more info.
"1" (y, associated measurements)
If X is composed of more than one array (e.g. connectivity matrix + node features), these are flattened to "0", "1", "2", and so on. y is ALWAYS the last one in that list (accessible via data.fields)

File: kinoml/ml/torch_models.py (hunk 2 above)
Comment (t-kimber) [reply kinoml/ml/torch_models.py:321]: Adapted docstrings accordingly.

File: kinoml/ml/torch_models.py (hunk 6 above)
Comment (t-kimber) [reply kinoml/ml/torch_models.py]: I changed it to protein for broader use cases in the future.

File: kinoml/ml/torch_models.py (hunk 7 above)
Comment (t-kimber) [reply kinoml/ml/torch_models.py]: Same here.

File: kinoml/ml/torch_models.py (hunk 5 above)
Comment (t-kimber) [reply kinoml/ml/torch_models.py]: You're right! Deleted it accordingly.

**Comment by t-kimber** [comment]
@schallerdavid @jchodera , i've added more detailed docstrings and hope it makes this a bit clearer. Hopefully it can be merged soon so that the experiments-binding-affinity repo doesn't get affected too much :)

**Comment by schallerdavid** [comment]
@t-kimber Looks good to me 👍

**Pull request merged**